cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -362,15 +362,25 @@ private void publishStartProgram(String runPayload, String peerName) throws IOEx
     Map<String, String> systemArgs = new HashMap<>(programOpts.getArguments().asMap());
     // Remove the plugin artifact archive argument from options and let the program runner recreate it
     systemArgs.remove(ProgramOptionConstants.PLUGIN_ARCHIVE);
-    systemArgs.put(ProgramOptionConstants.PEER_NAME, peerName);
+    systemArgs.put(ProgramOptionConstants.PEER_NAME, peerInfo.getName());
     systemArgs.put(ProgramOptionConstants.RUNTIME_NAMESPACE, message.getRuntimeNamespace());
     systemArgs.put(ProgramOptionConstants.PROGRAM_RESOURCE_URI, programDir.toURI().toString());
     systemArgs.put(ProgramOptionConstants.CLUSTER_MODE, ClusterMode.ISOLATED.name());
     SystemArguments.addProfileArgs(systemArgs, Profile.NATIVE);
     ProgramOptions updatedOpts = new SimpleProgramOptions(programId, new BasicArguments(systemArgs),
                                                           programOpts.getUserArguments());
     provisionerNotifier.provisioning(programRunId, updatedOpts, programDescriptor, "");
-    LOG.debug("Published program start message for run {}", programRunId.getRun());
+    LOG.debug("Published program start message for program run {}", programRunId);
+    // Check if the peer is allowed to run programs on the namespace specified in the launch message
+    // We first publish the PROVISIONING message to ensure that the run record is created.
+    if (peerInfo.getMetadata().getNamespaceAllocations().stream()
+      .noneMatch(n -> n.getNamespace().equals(message.getRuntimeNamespace()))) {
+      LOG.warn("Peer {} is not authorized run programs on namespace {}",
```

Comment:
run -> to run

Reply:
this sounds like a rbac thing when it's really about the connection namespaces being different. Would be good for the message to mention something about the namespace not being in the peer namespaces so that it's more clear that's the problem and not some missing auth permission.

Reply:
fixed error msg